File: reverse_integer.py
def reverse_integer(integer):
    output = ""
    string = str(integer)

    if string[0] == "-":
        isNegative = True
        string = string[1:]
    else:
        isNegative = False

    i = len(string) - 1
    while i >= 0:
        output += string[i]
        i -= 1

    if isNegative:
        if -int(output) < (~(1 << 31)):
            return 0
        else:
            return int(output)
    else:
        if int(output) > (1 << 31):
            return 0
        else:
            return int(output)

print(reverse_integer(123))
print(reverse_integer(100))
print(reverse_integer(-123))
print(reverse_integer(1534236469))
print(reverse_integer(-2147483648))

File: reverse_integer.rb
# Given a 32-bit signed integer, reverse digits of an integer.
#
# Example 1:
#
# Input: 123
# Output: 321
# Example 2:
#
# Input: -123
# Output: -321
# Example 3:
#
# Input: 120
# Output: 21

# Pop last digit by % 10. Push by * 10 and adding popped
# T: O(log10 n)
# S: O(1)
MIN_INT = -2**31
MAX_INT = 2**31-1

def reverse(x)
  previous = 0
  while x != 0
    popped = x.remainder(10)
    current = previous * 10 + popped
    return 0 if current > MAX_INT or current < MIN_INT
    previous = current
    # for negative numbers -12.3 rounds down to -13, which we dont want
    x = (x / 10.0).to_int
  end
  previous
end

puts reverse(123)
puts reverse(-123)
